net/xia/ppal_ether/main.c

```diff
@@ -81,6 +81,264 @@ static int local_dump_ether(struct fib_xid *fxid, struct fib_xid_table *xtbl,
+	u32 bucket;
+	int rc;
+
+	/* Check for errors in cfg. */
+	if (!cfg->xfc_dst || !cfg->xfc_odev || !cfg->xfc_lladdr 
+	            || cfg->xfc_lladdr_len != cfg->xfc_odev->addr_len)
+		return -EINVAL;
+
+	id = cfg->xfc_dst->xid_id;
+	dev = cfg->xfc_odev;
+
+	fxid = ether_rt_iops->fxid_find_lock(&bucket, xtbl, id);
+	if (!fxid) {
+		rc = -ENOENT;
+		goto unlock_bucket;
+	}
+	if (fxid->fx_table_id != XRTABLE_MAIN_INDEX) {
+		rc = -EINVAL;
+		goto unlock_bucket;
+	}
+	mether = fxid_mether(fxid);
+
+	if (!cmp_addr(mether->neigh_addr,cfg->xfc_lladdr,dev)) {
+		rc = -EINVAL;
+		goto unlock_bucket;
+	}
+
+	neigh_addr = mether->neigh_addr;
+	del_interface_addr(neigh_addr);
+	free_interface_addr(neigh_addr);
+
+	ether_rt_iops->fxid_rm_locked(&bucket, xtbl, fxid);
+	fxid_free(xtbl, fxid);
+	rc = 0;
+
+unlock_bucket:
+	ether_rt_iops->fib_unlock(xtbl, &bucket);
+	return rc;
+}
+
+static int main_dump_ether(struct fib_xid *fxid, struct fib_xid_table *xtbl,
+			 struct xip_ppal_ctx *ctx, struct sk_buff *skb,
+			 struct netlink_callback *cb)
+{
+	struct nlmsghdr *nlh;
+	u32 portid = NETLINK_CB(cb->skb).portid;
+	u32 seq = cb->nlh->nlmsg_seq;
+	struct rtmsg *rtm;
+	struct fib_xid_ether_main *mether = fxid_mether(fxid);
+	struct xia_xid dst;
+	struct nlattr *ha_attr;
+	struct rtnl_xia_ether_addrs *rtha;
+	struct interface_addr *pos_ia;
+
+	nlh = nlmsg_put(skb, portid, seq, RTM_NEWROUTE, sizeof(*rtm),
+			NLM_F_MULTI);
+	if (nlh == NULL)
+		return -EMSGSIZE;
+
+	rtm = nlmsg_data(nlh);
+	rtm->rtm_family = AF_XIA;
+	rtm->rtm_dst_len = sizeof(struct xia_xid);
+	rtm->rtm_src_len = 0;
+	rtm->rtm_tos = 0;
+	rtm->rtm_table = XRTABLE_MAIN_INDEX;
+	rtm->rtm_protocol = RTPROT_UNSPEC;
+	rtm->rtm_scope = RT_SCOPE_UNIVERSE;
+	rtm->rtm_type = RTN_UNICAST;
+	rtm->rtm_flags = 0;
+
+	dst.xid_type = xtbl->fxt_ppal_type;
+	memmove(dst.xid_id, fxid->fx_xid, XIA_XID_MAX);
+	if (unlikely(nla_put(skb, RTA_DST, sizeof(dst), &dst)))
+		goto nla_put_failure;
+
+	/* Hardware addresses. */
+	ha_attr = nla_nest_start(skb, RTA_MULTIPATH);
+	if (!ha_attr)
+		goto nla_put_failure;
+
+	rtha = nla_reserve_nohdr(skb, sizeof(*rtha));
+	if (!rtha)
+		goto nla_put_failure;
+
+	rcu_read_lock();
+	pos_ia = rcu_dereference(mether->neigh_addr);
+
+	rtha->interface_addr_len = pos_ia->outgress_interface->addr_len;
+	memmove(rtha->interface_addr, pos_ia->ha, rtha->interface_addr_len);
+	rtha->interface_index = pos_ia->outgress_interface->ifindex;
+	
+	rcu_read_unlock();
+	
+	/* No attributes. */
+
+	/* Length of rtnetlink header + attributes. */
+	rtha->attr_len = nlmsg_get_pos(skb) - (void *)rtha;
+
+	nla_nest_end(skb, ha_attr);
+
+	nlmsg_end(skb, nlh);
+	return 0;
+
+nla_put_failure:
+	nlmsg_cancel(skb, nlh);
+	return -EMSGSIZE;
+}
+
+/* Call using fxid_free only. */
+void main_free_ether(struct fib_xid_table *xtbl, struct fib_xid *fxid)
+{
+	struct fib_xid_ether_main *mether = fxid_mether(fxid);
+
+	xdst_free_anchor(&mether->xem_anchor);
+	mether->xem_dead = true;
+	mether_finish_destroy(mether);
```

Comment:
1. In the final code, this is the only place that calls `mether_finish_destroy()` and the body of `mether_finish_destroy()` is tiny (just two lines). It would be better to just move the body of `mether_finish_destroy()` here and drop `mether_finish_destroy()`.

2. Once `mether_finish_destroy()` is dropped, the field `xem_dead` of `struct fib_xid_ether_main` has no purpose, so we can drop this field as well.